pkg/pool-stable-phantom/contracts/StablePhantomPool.sol

```diff
@@ -851,19 +856,36 @@ contract StablePhantomPool is IRateProvider, BaseGeneralPool, StablePoolStorage,
         // However, a part of the invariant growth was due to non protocol swap fees (i.e. value accrued by the
         // LPs), so we only mint a percentage of this BPT amount: that which corresponds to protocol fees.
 
-        // We round down, favoring LP fees.
+        uint256 totalGrowthInvariant = StableMath._calculateInvariant(_getPostJoinExitAmp(), totalGrowthBalances);
+        uint256 swapGrowthInvariant = StableMath._calculateInvariant(_getPostJoinExitAmp(), swapGrowthBalances);
 
-        uint256 invariantRatio = preJoinInvariant.divDown(_postJoinExitInvariant);
-        uint256 protocolFeeAmount;
-        if (invariantRatio > FixedPoint.ONE) {
-            // This condition should always be met outside of rounding errors (for non-zero swap fees).
+        // Total Growth = Invariant with masked rates / last invariant: swap fees + non-exempt token yields
+        // Swap Fee Growth = Invariant with old rates / last invariant: swap fees alone
+        // Growth due to yield = Total Growth / Swap Fee Growth
+        //                     = Invariant with masked rates / Invariant with old rates.
 
-            protocolFeeAmount = getProtocolFeePercentageCache(ProtocolFeeType.SWAP).mulDown(
-                invariantRatio.sub(FixedPoint.ONE).mulDown(virtualSupply)
-            );
+        // If the "growth" is negative, set the ratio to ONE: multiplying by (ratio - 1) will then result in zero fees.
+        uint256 yieldGrowthRatio = totalGrowthInvariant > swapGrowthInvariant
+            ? totalGrowthInvariant.divDown(swapGrowthInvariant)
+            : FixedPoint.ONE;
```

Comment:
@rabmarut pointed out that this might mean charging extra fees in the future if the rate were to grow again after it fell. Not sure what his conclusion was

Reply:
Hm, we have a very simple solution for `WeightedPool`, but it won't work here. For reference, the weighted invariant can be trivially broken into two components: swaps and yield. So in `WeightedPool`, we calculate those two components completely independently and then collect yield fees only when the yield component exceeds its previous all-time high. But that independence is lost here with the yield portion being extracted from total growth. It won't be possible to track an all-time high yield.

I'm not sure if there is a solution to this for `StablePhantomPool`; we might have to accept this risk. Tokens with exceptionally volatile rates could always be exempted.

Reply:
At least this won't be an issue in bb-a-usd since linear rates are monotonically increasing

Reply:
It's probably also safe for the great majority of yield-bearing tokens, whose rates would only decrease due to a black swan event (which causes far more trouble than some excess protocol fees). So, it's an edge case, and there is a workaround for rates that are expected to be volatile: the pool creator can exempt those tokens. It's not perfect but it's probably the best we can do for now.